components/smfSupport.ts

```diff
@@ -0,0 +1,182 @@
+/*
+ *     The Peacock Project - a HITMAN server replacement.
+ *     Copyright (C) 2021-2023 The Peacock Project Team
+ *
+ *     This program is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU Affero General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU Affero General Public License for more details.
+ *
+ *     You should have received a copy of the GNU Affero General Public License
+ *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import { Controller } from "./controller"
+import { existsSync, readFileSync } from "fs"
+import { getFlag } from "./flags"
+import { log, LogLevel } from "./loggingInterop"
+import { MissionManifest, SMFLastDeploy } from "./types/types"
+import { basename, join } from "path"
+import { readFile } from "fs/promises"
+import { menuSystemDatabase } from "./menus/menuSystem"
+import { parse } from "json5"
+
+type LastServerSideData = SMFLastDeploy["lastServerSideStates"]
+
+export class SMFSupport {
+    public readonly lastDeploy: SMFLastDeploy
+
+    constructor(private readonly controller: Controller) {
+        const dataPath = SMFSupport.modFrameworkDataPath
+
+        if (dataPath && existsSync(dataPath)) {
+            this.lastDeploy = parse(readFileSync(dataPath).toString())
+            return
+        }
+
+        this.lastDeploy = null
+    }
+
+    static get modFrameworkDataPath() {
+        return (
+            (process.env.LOCALAPPDATA &&
+                join(
+                    process.env.LOCALAPPDATA,
+                    "Simple Mod Framework",
+                    "lastDeploy.json",
+                )) ||
+            false
+        )
+    }
+
+    private async executePlugin(plugin: string) {
+        if (!existsSync(plugin)) return
+
+        await this.controller.executePlugin(
```

Comment:
Is `this.controller` actually defined? The constructor doesn't seem to set it.

Reply:
Syntax sugar, if you put access modifiers on a constructor parameter, it automatically adds this.x = x

Reply:
Oh cool, didn't know that